<b>Using data create a model (Linear Regression) to predict a
house&#39;s value. We want to be able to understand what creates value
in a house, as though we were a real estate developer. The project
should follow the guideline as:</b>
1. Examine and explore data (visualization, interactions among
features)
2. Apply the model for prediction with holdout and cross
validation
3. Using PCA, apply the model with holdout and cross
validation
4. Visualize the residue and homoscedasticity
5. Tune the model if necessary
6. Write up analysis for each section (for example, explain why
the model is overfitting, explain why applying PCA is better,
etc.)
7. Include conclusions for summary

In [1]:
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols
from matplotlib import rc
import matplotlib.ticker as ticker
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
%matplotlib inline

# 1. Examine and explore data (visualization, interactions among features)

In [2]:
data=pd.read_csv("Melbourne_housing_FULL.csv")

In [3]:
data.shape

(34857, 21)

# There are 34857 rows and 21 columns. This sample is large enough for testing.

In [7]:
data.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

        Date  Distance  Postcode  Bedroom2  Bathroom  Car  Landsize  \
0  3/09/2016       2.5    3067.0       2.0       1.0  1.0     126.0   
1  3/12/2016       2.5    3067.0       2.0       1.0  1.0     202.0   
2  4/02/2016       2.5    3067.0       2.0       1.0  0.0     156.0   
3  4/02/2016       2.5    3067.0       3.0       2.0  1.0       0.0   
4  4/03/2017       2.5    3067.0       3.0       2.0  0.0     134.0   

   BuildingArea  YearBuilt         CouncilArea  Lattitude  Longtitude  \
0           NaN        NaN  Yarra City Council   -37.8014    144.9958   
1           NaN        NaN  Yarra City Council   -37.7996    144.9984   
2          79.0     1900.0  Yarra City Council   -37.8079    144.9934   
3           NaN        NaN  Yarra City Council   -37.8114    145.0116   
4         150.0     1900.0  Yarra City Council   -37.8093    144.9944   

              Regionname  Propertycount  
0  Northern Metropolitan         4019.0  
1  Northern Metropolitan         4019.0  
2  Northern Metropolitan         4019.0  
3  Northern Metropolitan         4019.0  
4  Northern Metropolitan         4019.0

In [15]:
pd.set_option('display.max_columns', 500,'display.max_rows', 500)

In [16]:
data.dtypes

Suburb            object
Address           object
Rooms              int64
Type              object
Price            float64
Method            object
SellerG           object
Date              object
Distance         float64
Postcode         float64
Bedroom2         float64
Bathroom         float64
Car              float64
Landsize         float64
BuildingArea     float64
YearBuilt        float64
CouncilArea       object
Lattitude        float64
Longtitude       float64
Regionname        object
Propertycount    float64
dtype: object

In [17]:
data.describe(include="all")

Suburb       Address         Rooms   Type         Price Method  \
count       34857         34857  34857.000000  34857  2.724700e+04  34857   
unique        351         34009           NaN      3           NaN      9   
top     Reservoir  5 Charles St           NaN      h           NaN      S   
freq          844             6           NaN  23980           NaN  19744   
mean          NaN           NaN      3.031012    NaN  1.050173e+06    NaN   
std           NaN           NaN      0.969933    NaN  6.414671e+05    NaN   
min           NaN           NaN      1.000000    NaN  8.500000e+04    NaN   
25%           NaN           NaN      2.000000    NaN  6.350000e+05    NaN   
50%           NaN           NaN      3.000000    NaN  8.700000e+05    NaN   
75%           NaN           NaN      4.000000    NaN  1.295000e+06    NaN   
max           NaN           NaN     16.000000    NaN  1.120000e+07    NaN   

       SellerG        Date      Distance      Postcode      Bedroom2  \
count    34857       34857  34856.000000  34856.000000  26640.000000   
unique     388          78           NaN           NaN           NaN   
top     Jellis  28/10/2017           NaN           NaN           NaN   
freq      3359        1119           NaN           NaN           NaN   
mean       NaN         NaN     11.184929   3116.062859      3.084647   
std        NaN         NaN      6.788892    109.023903      0.980690   
min        NaN         NaN      0.000000   3000.000000      0.000000   
25%        NaN         NaN      6.400000   3051.000000      2.000000   
50%        NaN         NaN     10.300000   3103.000000      3.000000   
75%        NaN         NaN     14.000000   3156.000000      4.000000   
max        NaN         NaN     48.100000   3978.000000     30.000000   

            Bathroom           Car       Landsize  BuildingArea     YearBuilt  \
count   26631.000000  26129.000000   23047.000000   13742.00000  15551.000000   
unique           NaN           NaN            NaN           NaN           NaN   
top              NaN           NaN            NaN           NaN           NaN   
freq             NaN           NaN            NaN           NaN           NaN   
mean        1.624798      1.728845     593.598993     160.25640   1965.289885   
std         0.724212      1.010771    3398.841946     401.26706     37.328178   
min         0.000000      0.000000       0.000000       0.00000   1196.000000   
25%         1.000000      1.000000     224.000000     102.00000   1940.000000   
50%         2.000000      2.000000     521.000000     136.00000   1970.000000   
75%         2.000000      2.000000     670.000000     188.00000   2000.000000   
max        12.000000     26.000000  433014.000000   44515.00000   2106.000000   

                    CouncilArea     Lattitude    Longtitude  \
count                     34854  26881.000000  26881.000000   
unique                       33           NaN           NaN   
top     Boroondara City Council           NaN           NaN   
freq                       3675           NaN           NaN   
mean                        NaN    -37.810634    145.001851   
std                         NaN      0.090279      0.120169   
min                         NaN    -38.190430    144.423790   
25%                         NaN    -37.862950    144.933500   
50%                         NaN    -37.807600    145.007800   
75%                         NaN    -37.754100    145.071900   
max                         NaN    -37.390200    145.526350   

                   Regionname  Propertycount  
count                   34854   34854.000000  
unique                      8            NaN  
top     Southern Metropolitan            NaN  
freq                    11836            NaN  
mean                      NaN    7572.888306  
std                       NaN    4428.090313  
min                       NaN      83.000000  
25%                       NaN    4385.000000  
50%                       NaN    6763.000000  
75%                  

In [18]:
data.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             7610
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
YearBuilt        19306
CouncilArea          3
Lattitude         7976
Longtitude        7976
Regionname           3
Propertycount        3
dtype: int64